### PR TITLE
Show battery charging state in the config panels

### PR DESCRIPTION
**Pull request by bdraco**
## Proposed change

The battery charging state is now checked when generating the battery icon.

Screen_Shot_2020-07-09_at_3_35_46_PM
Screen_Shot_2020-07-09_at_3_36_11_PM
Screen_Shot_2020-07-09_at_3_36_20_PM
Screen_Shot_2020-07-09_at_3_36_25_PM

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/common/entity/battery_icon.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  // hass:battery-charging-20
+  // hass:battery-charging-30
+  // hass:battery-charging-40
+  // hass:battery-charging-50
+  // hass:battery-charging-60
+  // hass:battery-charging-70
+  // hass:battery-charging-80
+  // hass:battery-charging-90
+  // hass:battery-outline
+  // hass:battery-alert
```
Comment (bramkragten) [inline src/common/entity/battery_icon.ts:49]: This is no longer needed/used

File: src/common/entity/battery_icon.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  const battery = Number(batteryState.state);
+  const battery_charging =
+    batteryChargingState && batteryChargingState.state === "on";
+
+  if (isNaN(battery)) {
+    return "hass:battery-unknown";
+  }
+
+  // We obscure 'hass' in iconname so this name does not get picked up
+  var icon = `${"hass"}:battery`;
```
Comment (bramkragten) [inline src/common/entity/battery_icon.ts:17]: Not needed anymore
```suggestion
  var icon = "hass:battery";
```

File: src/components/entity/ha-battery-icon.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import { html } from "@polymer/polymer/lib/utils/html-tag";
+/* eslint-plugin-disable lit */
+import { PolymerElement } from "@polymer/polymer/polymer-element";
+import { batteryIcon } from "../../common/entity/battery_icon";
+import "../ha-icon";
+
+class HaBatteryIcon extends PolymerElement {
```
Comment (bramkragten) [inline src/components/entity/ha-battery-icon.js:7]: I know you just copied this, but we should not add new PolymerElements, we are trying to get rid of them 😄

File: src/panels/config/devices/ha-config-devices-dashboard.ts [hunk 4]
```diff
@@ -167,7 +168,10 @@ export class HaConfigDeviceDashboard extends LitElement {
                 )
                 .join(", ")
             : "No integration",
-          battery_entity: this._batteryEntity(device.id, deviceEntityLookup),
+          battery_entity: [
```
Comment (bramkragten) [inline src/panels/config/devices/ha-config-devices-dashboard.ts:171]: Update the `DeviceRowData` interface to:
`battery_entity: [string | undefined, string | undefined];`

File: src/panels/config/devices/ha-config-devices-dashboard.ts [hunk 5]
```diff
@@ -201,17 +205,21 @@ export class HaConfigDeviceDashboard extends LitElement {
               sortable: true,
               type: "numeric",
               width: "90px",
-              template: (batteryEntity: string) => {
-                const battery = batteryEntity
-                  ? this.hass.states[batteryEntity]
+              template: (batteryEntityPair: Array<string>) => {
```
Comment (bramkragten) [inline src/panels/config/devices/ha-config-devices-dashboard.ts:208]:
```suggestion
              template: (batteryEntityPair: DeviceRowData["battery_entity"]) => {
```

File: src/components/entity/ha-battery-icon.js (hunk 3 above)
Comment (bdraco) [reply src/components/entity/ha-battery-icon.js:7]: What should I do instead here?

Comment (bramkragten) [reply src/components/entity/ha-battery-icon.js:7]: `LitElement`

Comment (bramkragten) [reply src/components/entity/ha-battery-icon.js:7]:
```ts
@customElement("ha-battery-icon")
class HaBatteryIcon extends LitElement {
    
  @property() public batteryStateObj;
  @property() public batteryChargingStateObj;

  protected render() {
    return html`
      <ha-icon
        .icon=${batteryIcon(batteryStateObj, batteryChargingStateObj)}
      ></ha-icon>
    `;
  }
}
```

Comment (bdraco) [reply src/components/entity/ha-battery-icon.js:7]: Thank you. So - much - cleaner.

Comment (balloob) [reply src/components/entity/ha-battery-icon.js:7]: Instead of using an `icon`, we should probably import the icon paths directly and use `ha-svg-icon` ?

Comment (bramkragten) [reply src/components/entity/ha-battery-icon.js:7]: Yes, but that would only work for this element so would need separate logic.

File: src/common/entity/battery_icon.ts (hunk 1 above)
Comment (bdraco) [reply src/common/entity/battery_icon.ts:49]: removed

File: src/components/entity/ha-battery-icon.js (hunk 3 above)
Comment (bdraco) [reply src/components/entity/ha-battery-icon.js]: `import { mdiBatteryCharging40, mdiBatteryCharging50, .... } from '@mdi/js';`

Seems like I'd need to import about 25 items and make an object to look them up in order to implement `ha-svg-icon`, unless there is a better way??

Comment (bramkragten) [reply src/components/entity/ha-battery-icon.js]: No, that would be the way.

The benefit of that is that we don't have to load an entire (or multiple) icon chunk of MDI icons and requests to the database where they are cached.

**Review by bramkragten — APPROVED** [review]

**Pull request merged**